src/VisualStudio/Core/Def/ProjectSystem/BrokeredService/WorkspaceProject.cs

```diff
@@ -0,0 +1,146 @@
+        // for a few edge cases. Once the project system has moved onto this directly, we can flatten the implementations out.
+        private readonly IWorkspaceProjectContext _project;
+
+        public WorkspaceProject(IWorkspaceProjectContext project)
+        {
+            _project = project;
+        }
+
+        public ValueTask DisposeAsync()
+        {
+            _project.Dispose();
+            return ValueTaskFactory.CompletedTask;
+        }
+
+        public async Task AddAdditionalFilesAsync(IReadOnlyList<string> additionalFilePaths, CancellationToken cancellationToken)
+        {
+            await using var batch = _project.CreateBatchScope().ConfigureAwait(false);
+
+            foreach (var additionalFilePath in additionalFilePaths)
+                _project.AddAdditionalFile(additionalFilePath);
+        }
+
+        public async Task RemoveAdditionalFilesAsync(IReadOnlyList<string> additionalFilePaths, CancellationToken cancellationToken)
+        {
+            await using var batch = _project.CreateBatchScope().ConfigureAwait(false);
+
+            foreach (var additionalFilePath in additionalFilePaths)
+                _project.RemoveAdditionalFile(additionalFilePath);
+        }
+
+        public async Task AddAnalyzerConfigFilesAsync(IReadOnlyList<string> analyzerConfigPaths, CancellationToken cancellationToken)
+        {
+            await using var batch = _project.CreateBatchScope().ConfigureAwait(false);
+
+            foreach (var analyzerConfigPath in analyzerConfigPaths)
+                _project.AddAnalyzerConfigFile(analyzerConfigPath);
+        }
+        public async Task RemoveAnalyzerConfigFilesAsync(IReadOnlyList<string> analyzerConfigPaths, CancellationToken cancellationToken)
+        {
+            await using var batch = _project.CreateBatchScope().ConfigureAwait(false);
+
+            foreach (var analyzerConfigPath in analyzerConfigPaths)
+                _project.RemoveAnalyzerConfigFile(analyzerConfigPath);
+        }
+
+        public async Task AddAnalyzerReferencesAsync(IReadOnlyList<string> analyzerPaths, CancellationToken cancellationToken)
+        {
+            await using var batch = _project.CreateBatchScope().ConfigureAwait(false);
+
+            foreach (var analyzerPath in analyzerPaths)
+                _project.AddAnalyzerReference(analyzerPath);
+        }
+
+        public async Task RemoveAnalyzerReferencesAsync(IReadOnlyList<string> analyzerPaths, CancellationToken cancellationToken)
+        {
+            await using var batch = _project.CreateBatchScope().ConfigureAwait(false);
+
+            foreach (var analyzerPath in analyzerPaths)
+                _project.RemoveAnalyzerReference(analyzerPath);
+        }
+
+        public async Task AddMetadataReferencesAsync(IReadOnlyList<MetadataReferenceInfo> metadataReferences, CancellationToken cancellationToken)
```

Comment:
why IReadOnlyList out of curiosity?

Reply:
This was explicitly mentioned in the [RPC contracts best practices](https://dev.azure.com/devdiv/DevDiv/_git/VS.RPC.Contracts?anchor=version-compatibility-concerns&path=%2Fdoc%2Fbestpractices.md&_a=preview):

> Use readonly collections interfaces in RPC method signatures (e.g. IReadOnlyList<T>) rather than concrete types (e.g. List<T> or T[]). This allows for potentially more efficient deserialization.

Maybe @AArnott has further thoughts; but the downside of a strongly typed thing here is we're then stuck with that. Also because they're abstract means they don't technically even need an immutable implementation if that has extra memory overhead.

Reply:
Besides interfaces leaving the deserializer to do the most efficient thing, using a *read only* interface makes it more clear to the receiver that they were not given some data structure that they can mutate and expect to be mutating the sender's copy. This guidance applies to more than just collections. We encourage *every* data type to be immutable so there is never any question that the receiver cannot change it and expect the sender to change as a result.

Reply:
Sorry, I meant: why IROL instead on ImmutableArray.

Reply:
That's up to you. But a reason not to use `ImmutableArray<T>` in a deserializer is that if the deserializer doesn't know the length ahead of time, it means an extra copy and allocation of the collection is required.
Now it so happens that if you're using MessagePack as your formatter, it _does_ know the length ahead of time and leverages that to use the optimal builder+MoveToImmutable pattern, so that could serve you well here.